sql/mysqld.cc

```diff
@@ -5587,6 +5591,70 @@ static void test_lc_time_sz()
 }
 #endif//DBUG_OFF
 
+static void check_upgrade(void)
+{
+  File info_file;
+  char upgrade_from_version[1024]= {0};
+  size_t length;
+  ulong version_id;
+
+  uint major, minor, version;
+  char *point= upgrade_from_version, *end_point, *pos;
+
+  DBUG_ENTER("check_upgrade");
+
+  if (upgrade_info_file != 0)
+  {
+    fn_format(upgrade_info_file, "mysql_upgrade_info", mysql_real_data_home_ptr, "", MYF(0));
+    DBUG_PRINT("info", ("mysql_upgrade_info path: %s", mysql_real_data_home_ptr));
+  }
+
+  if ((info_file= my_open(upgrade_info_file, O_RDONLY, MYF(0))) < 0)
+  {
+    sql_print_information("Cannot open mysql_upgrade_info file: %s. Please run mariadb-upgrade.", upgrade_info_file);
+    DBUG_VOID_RETURN;
+  }
+
+  if (my_lock(info_file, F_RDLCK, 0, 1, MYF(0)))
+    goto end; // If Another process is holding a write lock avoid checking
+
+  (void) my_seek(info_file, 0, SEEK_SET, MYF(0));
+  /* We have -3 here to make conversion below safe (similar to mysql_upgrade) */
```

Comment:
if you've got more full explanation for the -3 can this be used.

Reply:
I took that from `mysql_upgrade.c`:
https://github.com/MariaDB/server/blob/6252a281b5020d28685777c11ee27688a9db3d66/client/mysql_upgrade.c#L820

My understanding is that the -3 leave at least 3 zeroes in `upgrade_from_version` to ensure that the 3 following calls to `strtoul` will succeed always. But I might be understanding this wrong.

Reply:
I wondered about this too, and I think your understanding is correct, @christian7007.

(Probably a good idea to improve the comments in _both_ places :-D)

Reply:
thanks for checking.